File: 0074-Search-a-2D-Matrix/lc0074.py
# Python solution.

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = bisect.bisect(matrix, [target+0.5])
        if m:
            n = bisect.bisect(matrix[m-1], target)
            return len(matrix[0]) > 0 and matrix[m-1][n-1] == target
        return False

File: 0074-Search-a-2D-Matrix/lc0074.js
// JavaScript solution.

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    let m = matrix.length;
    if (m > 0) {
        let n = matrix[0].length;
        let lo = 0, hi = m * n;
        while (lo < hi) {
            let mid = (lo + hi) >> 1;
            let x = matrix[Math.trunc(mid/n)][mid%n];
            if (x < target) { lo = mid+1; }
            else if (x > target) { hi = mid; }
            else { return true; }
        }
    }
    return false;
};
